.github/workflows/rust-test.yml

```diff
@@ -68,20 +72,72 @@ env:
   CARGO_HTTP_USER_AGENT: "shipyard ${{ secrets.CARGO_PRIVATE_REGISTRY_TOKEN }}"
 
 jobs:
+  build-matrix:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+
+      - name: Build Test Matrix
+        id: build
+        shell: bash
+        run: |
+          workspaces="${{ inputs.workspaces }}"
+          workspace_list=($(echo -n ${workspaces:1:${#workspaces}-2} | tr , " "))
+
+          if [ "${{ github.event_name }}" != "pull_request" ]; then
+            echo "matrix=$workspaces" >> $GITHUB_OUTPUT
+            exit 0
+          fi
+
+          git fetch origin ${{ github.base_ref }}
+
+          output_paths=''
+          root_pwd="$PWD"
+
+          for workspace in "${workspace_list[@]}"; do
+            cd "$root_pwd"
+            cd "$workspace"
```

Comment:
Shouldn't we use `pushd` `popd` instead?

Reply:
An earlier version used them, yes. This variant is more stable against accidentally forgetting a popd when you break or continue the loop and also is easier on your shell if you run the script locally. If you feel strongly about it, then I can change it back.